### ISSUE TITLE
GCC-13/14 bottle not working with newer MacOSX SDKs

### ISSUE THREAD

**Issue by anddebol** [Homebrew/homebrew-core#194778]
### `brew gist-logs <formula>` link OR `brew config` AND `brew doctor` output

```shell
No logs.
```

### Verification

- [X] My `brew doctor` output says `Your system is ready to brew.` and am still able to reproduce my issue.
- [X] I ran `brew update` and am still able to reproduce my issue.
- [X] I have resolved all warnings from `brew doctor` and that did not fix my problem.
- [X] I searched for recent similar issues at https://github.com/Homebrew/homebrew-core/issues?q=is%3Aissue and found no duplicates.

### What were you trying to do (and why)?

Updating CLT to 16 will break gcc includes 

### What happened (include all command output)?

since latest gcc bottle is not built vs. CLT 16.0  it compiling with it will produce stdio.h errors similar to [this issue](https://github.com/iains/gcc-14-branch/issues/14) ([similar issue](https://github.com/iains/gcc-14-branch/issues/12))

### What did you expect to happen?

Bottle is updated to use (https://github.com/iains/gcc-14-branch/releases/tag/gcc-14.2-darwin-r2)

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
(assuming you have MacOSX15.sdk as current)
any c++ code which includes stdio.h
```

**Comment by carlocab** [comment]
Works for me:
```
❯ cat <<C | gcc-14 --sysroot=/Library/Developer/CommandLineTools/SDKs/MacOSX15.sdk -xc - -o test
#include <stdio.h>
int main(void) {
  printf("test\n");
  return 0;
}
C

❯ ./test
test
```

What's the output of your `brew config` and `brew doctor`?

**Comment by Bo98** [comment]
r2 release only just came out. We will update GCC 14 to use it soon and build a macOS 15 bottle.

GCC 13 is probably further away.

**Comment by saschasc** [comment]
@Bo98 Is somewhere described how the [patch file](https://github.com/Homebrew/homebrew-core/blob/3a67f18a7e886c897298971adef3990c08669f75/Formula/g/gcc.rb#L15) can be created? I was testing/helping Iains and it would be easier for me to adapt the Homebrew recipe.

**Comment by saschasc** [comment]
@anddebol Which OS are you running the tests?

**Comment by Bo98** [comment]
> @Bo98 Is somewhere described how the [patch file](https://github.com/Homebrew/homebrew-core/blob/3a67f18a7e886c897298971adef3990c08669f75/Formula/g/gcc.rb#L15) can be created?

@fxcoudert might have the command. On a basic level it's a diff between `gcc-14.2-darwin-r*` and `releases/gcc-14.2.0` with I think markdown files and certain testsuite-only commits removed to stripped from the diff. Will need to check with FX exactly how it's constructed.

**Comment by fxcoudert** [comment]
Also importantly, we will need to rebuild the new bottles when CI machines have been updated to latest Xcode on each macOS version.

The diff file is as @Bo98 explained: I remove some readme files, testsuite changes, and other things that are irrelevant to homebrew (to keep the size minimal).

**Comment by anddebol** [comment]
@carlocab
> Works for me:

It doesn't work (stdio errors) for me with --sysroot=/Library/Developer/CommandLineTools/SDKs/MacOSX15.sdk (which is now is default on my system)
it does work with --sysroot=/Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk though

Maybe it is Intel and not AppleSilicon issue?

> What's the output of your `brew config` and `brew doctor`?

brew config:
```
brew config
HOMEBREW_VERSION: 4.4.1-46-g1fdbc6f
ORIGIN: https://github.com/Homebrew/brew
HEAD: 1fdbc6fe5a1e3700aea3ac1d167c90a19782a2b8
Last commit: 66 minutes ago
Core tap HEAD: 3a67f18a7e886c897298971adef3990c08669f75
Core tap last commit: 80 minutes ago
Core tap JSON: 17 Oct 16:00 UTC
Core cask tap HEAD: 37788e070ae3e8091770c46515ff64b86043468a
Core cask tap last commit: 61 minutes ago
Core cask tap JSON: 17 Oct 16:00 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.pv9EdHLio6/org.xquartz:0
HOMEBREW_MAKE_JOBS: 16
HOMEBREW_SORBET_RUNTIME: set
Homebrew Ruby: 3.3.5 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/3.3.5/bin/ruby
CPU: 16-core 64-bit kabylake
Clang: 16.0.0 build 1600
Git: 2.47.0 => /usr/local/bin/git
Curl: 8.7.1 => /usr/bin/curl
macOS: 14.7-x86_64
CLT: 16.0.0.0.1.1724870825
Xcode: 16.0
```
brew doctor:
```
Your system is ready to brew.
```

**Comment by fxcoudert** [comment]
GCC does not guarantee compatibility between major macOS SDK versions. You need to use the SDK that it was built against. Other configurations are (sadly) not always working, because Apple introduces incompatible changes in their SDK headers.

**Comment by anddebol** [comment]
Thanks, @fxcoudert!

**Comment by Bo98** [comment]
GCC (or at least the way we build it) should default to the SDK it was built against so you shouldn't need to explicitly set the SDK as long as you have the Command Line Tools installed.

**Comment by saschasc** [comment]
@Bo98 You are right. I still have an issue, but this doesn't seem to be gcc related, but eventually CMake and/or Conan related. When I've installed both XCT 15 and 16 on macOS Sonoma it seems that the sysroot gets overridden somehow and I must explicitly set the `SDKROOT` variable to the correct path.

Why does e.g. the following command on macOS 14 show to the macOS 15.x SDK and not the 14.x SDK directory?

```
~ % xcrun -sdk macosx --show-sdk-path
/Library/Developer/CommandLineTools/SDKs/MacOSX15.1.sdk
```

**Comment by carlocab** [comment]
Could be a CMake bug. Try passing 
```
-DCMAKE_OSX_SYSROOT=/Library/Developer/CommandLineTools/MacOSX14.sdk
```
to CMake.

As for what `xcrun` is doing, I suggest trying
```
unset DEVELOPER_DIR SDKROOT
xcrun --kill-cache
```
and see if that helps.

**Comment by Bo98** [comment]
For CMake, the default is it chooses the latest SDK, so either set the `SDKROOT` or `-DCMAKE_OSX_SYSROOT`.

The correct SDK can be subjective so tools using the latest aren't necessarily wrong - but can be risky for certain use cases. If you are writing in Swift using only Apple Frameworks and not low-level APIs, the latest SDK is often desirable and the best choice as it supports availability attributes, gives errors on APIs that Apple have decided they want to stop using even on older macOS (e.g. in macOS 15 they obsoleted the old screen sharing API and _really_ don't want you to use it anymore even on macOS 14) and it allows you to do things like `if #available(OSX 15.0, *) {` to enable new macOS 15 features. The moment however you touch low-level APIs, this is very harmful. If you use `libcurl` from a newer SDK, there's a chance your program may crash at runtime if you do compile-time feature/version checking.

This is sort of why Xcode.app only ships with the macOS 15 SDK, but the Command Line Tools ships with both macOS 14 SDK and macOS 15 SDK. If you're developing in Xcode.app, you're probably making a macOS app. For everything else, you're probably aware of the command line anyway.

For Homebrew, we always use the host SDK for everything except for software written in Swift and other software that only works with the full Xcode.app (e.g. things using `xcodebuild`) where we use the latest SDK. If in doubt, the SDK matching the running OS is the safest option.

`xcrun` is unfortunately very confusing. First of all, it depends on `xcode-select -p` which for many is Xcode.app. You need to set `DEVELOPER_DIR=/Library/Developer/CommandLineTools` for CLT SDKs to be picked.

Secondly, `xcrun` with and without `--sdk macosx` changes behaviour.

* `--sdk macosx` will select the latest SDK.
* Omitting it and just doing `xcrun --show-sdk-path` will select the system/host SDK, or the latest SDK if it cannot find the system SDK.

However, the behaviour of selecting the system/host SDK is broken when it comes to minor versions. The latest SDK for macOS 14 is 14.5, but the latest OS version for macOS 14 is 14.7. If you run 14.7, `xcrun --show-sdk-path` will try to find `MacOSX14.7.sdk` before giving up and returning the latest SDK instead. This is obviously totally incorrect.

What about `libxcselect`'s `xcselect_host_sdk_path` (https://developer.apple.com/documentation/xcselect/xcselect_host_sdk_path?language=objc)? Don't even bother. The host selection _still_ only checks for SDKs starting with `MacOSX10`.

**TLDR**: don't bother trying to get it to work and just do `/Library/Developer/CommandLineTools/SDKs/MacOSX$(sw_vers -productVersion | cut -d'.' -f1).sdk`, potentially falling back to `xcrun`/`libxcselect` if the CLT isn't installed and using a newer SDK is acceptable for your use case.

If you're embedding the SDK into a build (e.g. you are building GCC itself), the CLT's `MacOSX14.sdk` symlink is beneficial over `MacOSX14.5.sdk` anyway as it's stable across minor versions.

**Comment by carlocab** [comment]
> What about `libxcselect`'s `xcselect_host_sdk_path` ([developer.apple.com/documentation/xcselect/xcselect_host_sdk_path?language=objc](https://developer.apple.com/documentation/xcselect/xcselect_host_sdk_path?language=objc))? Don't even bother. The host selection _still_ only checks for SDKs starting with `MacOSX10`.

They might have finally fixed this...
```
❯ cat <<'C' | clang -xc - -lxcselect -o test; ./test
#include <xcselect.h>
#include <stdio.h>
int main(void) {
    char *path = NULL;
    xcselect_host_sdk_path(XCSELECT_HOST_SDK_POLICY_MATCHING_PREFERRED, &path);
    printf("%s\n", path);
    return 0;
}
C
/Library/Developer/CommandLineTools/SDKs/MacOSX15.0.sdk
```

**Comment by Bo98** [comment]
Haven't checked macOS 15, so will be nice if they have.

You want to check `XCSELECT_HOST_SDK_POLICY_MATCHING_ONLY` however. `PREFERRED` will fallback to latest.

**Comment by carlocab** [comment]
```
❯ cat <<'C' | /usr/bin/clang -xc - -lxcselect -o test; ./test
#include <xcselect.h>
#include <stdio.h>
int main(void) {
    char *path = NULL;
    xcselect_host_sdk_path(XCSELECT_HOST_SDK_POLICY_MATCHING_ONLY, &path);
    printf("%s\n", path);
    return 0;
}
C
/Library/Developer/CommandLineTools/SDKs/MacOSX15.0.sdk
```
But I didn't check the return value of `xcselect_host_sdk_path` so this may just well be UB.

**Comment by anddebol** [comment]
>I assume you are still running on macOS Sonoma?

@saschasc yes. I figured to use proper systemroot for CMAKE, thanks for the help @Bo98 and @carlocab 
Was totally unaware of any of this since I mostly use CLion.

**Comment by anddebol** [comment]
I also experience some bugs with ldd (not linking shared hdf5_hl and netcdf properly), and a bit afraid to update to Sequoia now. Can the miss match of OS/SDK version be also a culprit, and I actually need to update the OS?  What would be and a safer option?

**Comment by vszakats** [comment]
I've spent a few weeks earlier this year on this (and other) macOS build issues with non-Apple-clang compilers. The context was curl builds.

Can confirm the issue mentioned here, and I could trace back the root cause not to gcc itself, but the hack-layer applied on top. The hack layer is designed to override certain SDK headers. They are generated for a specific SDK version at built time, for the SDK found on the build machine.

The SDK version found on the user machine may be a different one. For GHA macos images, the default SDK turned out to be different in most cases. It can also be different for various other reasons, e.g. after an Xcode upgrade, while keeping using the previous Homebrew binaries, which get updates on a different cycle.

Either way, when the hack-layer build-time SDK differs from the actual SDK on the user machine,
instead of fixing issues, the hack-layer breaks the build in various, difficult to figure out ways.

The only workaround I've found is to selectively zap the hack-layer headers to avoid them kicking
in. An ugly hack and likely not applicable in all (or most) cases.

The stable solution would be (IMO) if the hack layer would be provided for each supported SDK 
and selected dynamically to match the selected one on the user machine.

There is more info about it here: https://github.com/curl/curl/pull/14097 (section 1.)
and there are links to all other issues found in this process. There is also a table 
showing the different combinations and the experienced fallouts.

After these, curl be built in any combination of xcode, SDK, gcc and llvm version:
https://github.com/curl/curl/blob/cadfe0de68355f1e5c90676ab03992c1afbf71c5/.github/workflows/macos.yml#L356-L590

Section 2 is about the `availability` fallout, which was fixed by updating the gcc patch
in 14.2.0_1. Thanks for that! The local workaround is still necessary for 12.4.0 and 13.3.0:
https://github.com/curl/curl/pull/15508

**Comment by Bo98** [comment]
> The SDK version found on the user machine may be a different one. For GHA macos images,
> the default SDK turned out to be different in most cases. It can also be different for other
> reasons, e.g. after an Xcode upgrade, while keeping using the previous Homebrew binaries,
> which get updates on a different cycle.

Homebrew GCC should generally work by default on most systems, and definitely on GitHub Actions, but only if you don't override the default via `-isysroot`. We set the default to `/Library/Developer/CommandLineTools/SDKs/MacOSX??.sdk` where `??` matches the major macOS version GCC was built on. If you remember to do `brew reinstall/upgrade gcc` when doing a major macOS upgrade on your local system, this path will always exist if you have the CLT installed correctly. It will always exist on GitHub Actions images.

For autotools, this generally should work out of the box. I don't think autotools messes with sysroots.

CMake unfortunately does mess with sysroots however, and usually does so in a way that breaks the SDK rule with GCC. When using CMake, it is strongly recommended you pass `-DCMAKE_OSX_SYSROOT=$(gcc-14 -print-sysroot)`. This arguably should be the default with GCC upstream.

> The stable solution would be (IMO) if the hack layer would be provided for each supported SDK and selected dynamically to match the selected one on the user machine.

Even this is dynamic. The `/Library/Developer/CommandLineTools/SDKs/MacOSX??.sdk` mentioned earlier is the _only_ SDK guaranteed to exist for a compatible CLT install (and no fixed SDK is guaranteed to exist for Xcode.app).

To follow the macOS 14 cycle:

Xcode 15.*:
-  CLT had `MacOSX13.sdk` and `MacOSX14.sdk`
-  Xcode.app had `MacOSX14.*.sdk` (minor version changing every couple months throughout the year) symlinked as `MacOSX.sdk` (which isn't stable for GCC as the hack headers vary depending on the major SDK version as Apple ship breaking changes on major releases).

Xcode 16.*:
- CLT had `MacOSX14.sdk` and `MacOSX15.sdk`
- Xcode.app had `MacOSX15.*.sdk` symlinked as `MacOSX.sdk`

So it's hard to support all of 13-15 at the same time as it needs to read the SDK at build time (for copyright reasons).

It however does suck - I agree and everyone upstream agrees too. The long term fix here is to just reduce the need of hack headers altogether. The improvements in `availability` handling is one such example of this - it wasn't a hack header fix but a new core GCC feature that removed the need for hacks.

> Section 2 is about the `availability` fallout, which was fixed by updating the gcc patch in 14.2.0_1. Thanks for that! The local workaround is still necessary for 12.4.0 and 13.3.0: [curl/curl#15508](https://github.com/curl/curl/pull/15508)

Yeah they will be back ported over the coming weeks.

Note that you can detect a working fix via: `#if defined(__has_feature) && __has_feature(attribute_availability)`

**Comment by Bo98** [comment]
Oh and to clarify: the preinstalled GCC version on GitHub Actions should be 14.2.0_1 fairly soon. `macos-14` version 20241106 is rolling out now and other versions will follow soon.

**Comment by vszakats** [comment]
> > The SDK version found on the user machine may be a different one. For GHA macos images,
> > the default SDK turned out to be different in most cases. It can also be different for other
> > reasons, e.g. after an Xcode upgrade, while keeping using the previous Homebrew binaries,
> > which get updates on a different cycle.
> 
> Homebrew GCC should generally work by default on most systems, and definitely on GitHub Actions, but only if you don't override the default via `-isysroot`. We set the default to `/Library/Developer/CommandLineTools/SDKs/MacOSX??.sdk` where `??` matches the major macOS version GCC was built on. If you remember to do `brew reinstall/upgrade gcc` when doing a major macOS upgrade on your local system, this path will always exist if you have the CLT installed correctly. It will always exist on GitHub Actions images.

I cannot confirm this for GHA runners. When doing this in the summer, out
of 12 combinations of gcc versions and default GHA SDKs only 4 had matching 
SDKs, and all the 8 mismatches failed out of the box. Interestingly even one
matching combination failed. This is 3 working combos out of 12.
(The complete table is the linked PR message.)

After the Seqouia upgrade there was a new macos-15 machine where the
latest gcc was the one built one OS release prior, with no newer release
available. Making a mismatch, and a build failure.

> For autotools, this generally should work out of the box. I don't think autotools messes with sysroots, unless you are doing so yourself.

I did all tests with both autotools and cmake. It's not trivial to get them
configured correctly, but once done, they produced the same results 
WRT to the matching or not-matching SDKs.

> CMake unfortunately does mess with sysroots however, and usually does so in a way that breaks the SDK rule with GCC. When using CMake, it is strongly recommended you pass `-DCMAKE_OSX_SYSROOT=$(gcc-14 -print-sysroot)`. This arguably should be CMake's default with GCC upstream.

Well, yes, but that makes GCC dictate the SDK, meaning to override the default
one. If the GCC-dictated version is actually offered by the runner. This isn't always
the case. And when it is the case, this is isn't easily discoverable and many users
don't do it. It also doesn't help that there are several interdependent issues at hand,
and untangling them can take quite a bit of effort. It's basically an undocumented
limitation when using GCC, which is IMO extremely difficult to discover.

> Even this is dynamic. The `/Library/Developer/CommandLineTools/SDKs/MacOSX??.sdk` mentioned earlier is the _only_ SDK guaranteed to exist for a compatible CLT install (and no fixed SDK is guaranteed to exist for Xcode.app).

Perhaps dynamic wasn't a good choice of word. What I mean is that GCC comes
with the hack layer generated from the locally installed SDK with patches applied
to it and GCC _build-time_.

Currently the only safe way to use these GCC builds is to have the very same 
SDK installed locally.

By "dynamic" I mean a solution where GCC isn't bound to a single SDK at GCC
build-time, but adapts to whatever installed SDK there is (within limits) at runtime
(use time?) and sets up the hack layer for that. I understand this is technically
tedious or even impossible, and a completely different system.

Ideally GCC should be able to handle the SDK without hack layers, but that's
probably impossible for different reasons. Though like with 14.2.0_1, if the 
hack layer can fix the SDK incompatibilites by updating the compiler instead of
offering header overlays, it's also better, but possibly (much) harder.

> So it's hard to support all of 13-15 at the same time as it needs to read the SDK at build time (for copyright reasons).

Yep, I totally get it. It's not easy.

> It however does suck - I agree and everyone upstream agrees too. The long term fix here is to just reduce the need of hack headers altogether. The improvements in `availability` handling is one such example of this - it wasn't a hack header fix but a new core GCC feature that removed the need for hacks.

Agreed!

> > Section 2 is about the `availability` fallout, which was fixed by updating the gcc patch in 14.2.0_1. Thanks for that! The local workaround is still necessary for 12.4.0 and 13.3.0: [curl/curl#15508](https://github.com/curl/curl/pull/15508)
> 
> Yeah they will be back ported over the coming weeks.
> 
> Note that you can detect a working fix via: `#if defined(__has_feature) && __has_feature(attribute_availability)`

Thanks! I cannot personally make tests with this locally (I'm stuck
at Monterey) but will try to give it a spin in CI.

> However, if you adjust sysroot handling as earlier, this should only be an issue on macOS 15 and we don't have older GCC built for macOS 15 intentionally for that reason so `brew install gcc@13` should still be using the macOS 14 SDKs by default for now (which will still exist on macOS 15 images until Xcode CLT 17). When the fix lands for GCC 13 we'll do a proper build of GCC 13 for macOS 15.

It's not easy. I'm trying to track the ever changing Xcode versions, SDK
versions and their defaults across 3 runners. Plus which gcc is preinstalled
on which image and what is their default SDK. Any of this can (and does)
change overnight (or with brew update). It's maintenance heavy and fragile.

I will consider changing the approach to forget the idea of "combinations" for
gcc, and lock it to known-good ones to make it easier.

Though I prefer using the bottles shipping with the images by default. To
keep the CI fast and lean and to limit the amount of env changes.

The current solution of zapping the hack-layer only works in CI anyway, and
will not fix the problem of anyone trying to do these builds locally. Thus its
value is somewhat academic.

> The above `__has_feature` check will however help detect older GCC 14 that we did ship a macOS 15 bottle for.

Appreciated!

Also thanks all the work you are doing with this.

**Comment by Bo98** [comment]
> Well, yes, but that makes GCC dictate the SDK, meaning to override the default
> one. If the GCC-dictated version is actually offered by the runner. This isn't always
> the case. And when it is the case, this is isn't easily discoverable and many users
> don't do it. It also doesn't help that there are several interdependent issues at hand,
> and untangling them can take quite a bit of effort. It's basically an undocumented
> limitation when using GCC, which is IMO extremely difficult to discover.

"Default" is largely subjective. There's three "defaults": https://developer.apple.com/documentation/xcselect/hostsdkpolicy. GCC follows the `matchingOnly` policy, albeit at build time rather than runtime.

If you mean CMake's default, sure. CMake has its own SDK selection logic that equates to `latest`.

For GCC, the default means: don't select the SDK yourself and let GCC decide. Ideally `-isysroot` would work - I agree. But that's a limitation of GCC on macOS and if you fight against that limitation you are signing up to dealing with horrible hacks every year when new breaking changes are made in future macOS SDKs and GCC hasn't been patched yet. Apple unfortunately simply don't test new SDKs against GCC so GCC ends up playing catch up.

> It's not easy. I'm trying to track the ever changing Xcode versions, SDK
> versions and their defaults across 3 runners. Plus which gcc is preinstalled
> on which image and what is their default SDK. Any of this can (and does)
> change overnight (or with brew update). It's maintenance heavy and fragile.

Using only the sysroot/SDK supported by GCC should reduce maintenance burden significantly.

If you are using the sysroot/SDK supported by GCC and still find you are needing header hacks - please let me know and I am happy to look into it further.

**Comment by Bo98** [comment]
> Note that you can detect a working fix via: `#if defined(__has_feature) && __has_feature(attribute_availability)`

Just to correct this, this should be:

```c
#ifdef __has_feature
#if __has_feature(attribute_availability)
...
```

as the preprocessor doesn't short-circuit parsing and you would get `missing binary operator before token "("`.

**Comment by vszakats** [comment]
> "Default" is largely subjective. There's three "defaults": https://developer.apple.com/documentation/xcselect/hostsdkpolicy. GCC follows the `matchingOnly` policy, albeit at build time rather than runtime.
> 
> If you mean CMake's default, sure. CMake has its own SDK selection logic that equates to `latest`.

I mean the one selected on GHA as default according to:
https://github.com/actions/runner-images/blob/main/images/macos/macos-15-arm64-Readme.md#xcode
and these:
```
xcrun --sdk macosx --show-sdk-path
xcrun --sdk macosx --show-sdk-version
```

(This might match with what CMake considers default.)

> For GCC, the default means: don't select the SDK yourself and let GCC decide. Ideally `-isysroot` would work - I agree. But that's a limitation of GCC on macOS and if you fight against that limitation you are signing up to dealing with horrible hacks every year when new breaking changes are made in future macOS SDKs and GCC hasn't been patched yet. Apple unfortunately simply don't test new SDKs against GCC so GCC ends up playing catch up.

Yeah, that's palpable. Even mainline llvm cannot always follow quickly
enough what Apple is doing (ref `__ENVIRONMENT_OS_VERSION_MIN_REQUIRED__`).

I'm not sure where would be the best way to document this, but it
could save a lot of time knowing that and making sure everyone has
_that_ SDK installed and points their build tool to it too when using gcc.

The gcc hack-layet itself could maybe print a warning to this effect,
if it's cleanly detectable.

> Using only the sysroot/SDK supported by GCC should reduce maintenance burden significantly.
> 
> If you are using the sysroot/SDK supported by GCC and still find you are needing header hacks - please let me know and I am happy to look into it further.

Thank you! I'll make tests and see how it goes.

**Comment by carlocab** [comment]
> CMake unfortunately does mess with sysroots however, and usually does so in a way that breaks the SDK rule with GCC. 

Upstream might fix this soon: https://gitlab.kitware.com/cmake/cmake/-/merge_requests/9977

**Comment by fxcoudert** [comment]
> Ideally GCC should be able to handle the SDK without hack layers, but that's
probably impossible for different reasons.

The root reason for the issue is: Apple's SDK often contain nonstandard extensions, without proper guards. We report them to Apple regularly, and they (most often) get fixed over time.

On all modern OS, GCC does not have to fix system headers/SDK.

**Comment by vszakats** [comment]
> as the preprocessor doesn't short-circuit parsing and you would get `missing binary operator before token "("` if you had the `defined` check within the same `#if` and ran with old GCC

Run into it, also with MSVC. Fixed them by splitting off the two conditions,
but indeed that would still fail with old Apple gcc (not CI tested).
I need to check for the reverse, making the clean iteration somewhat awkward:
```c
#if defined(__APPLE__) && !defined(__clang__) && defined(__GNUC__) && \
  defined(__has_attribute)
#  if !defined(__has_feature)
#    define availability curl_pp_attribute_disabled
#  elif !__has_feature(attribute_availability)
#    define availability curl_pp_attribute_disabled
#  endif
#endif
```

**Comment by vszakats** [comment]
Finished the test with locking the SDK to `gcc --print-sysroot` and it's a success:
https://github.com/curl/curl/actions/runs/11734032254/job/32690076166?pr=15518
I plan to merge it and drop the workaround used till now.

(Surprise discovery that the CLT comes with a different set of SDKs than Xcode.
They are not listed in the macos runner README.)

Thanks for the help to all!

**Comment by vszakats** [comment]
> > Ideally GCC should be able to handle the SDK without hack layers, but that's
> > probably impossible for different reasons.
> 
> The root reason for the issue is: Apple's SDK often contain nonstandard extensions, without proper guards. We report them to Apple regularly, and they (most often) get fixed over time.

Good to know and thanks for this effort too!

> On all modern OS, GCC does not have to fix system headers/SDK.

gcc-12 built with SDK 14 ships with these overlays in `include-fixed`:
```
  /opt/homebrew/Cellar/gcc@12/12.4.0/bin/../lib/gcc/12/gcc/aarch64-apple-darwin23/12/include-fixed/dispatch/object.h
  /opt/homebrew/Cellar/gcc@12/12.4.0/bin/../lib/gcc/12/gcc/aarch64-apple-darwin23/12/include-fixed/limits.h
  /opt/homebrew/Cellar/gcc@12/12.4.0/bin/../lib/gcc/12/gcc/aarch64-apple-darwin23/12/include-fixed/math.h
  /opt/homebrew/Cellar/gcc@12/12.4.0/bin/../lib/gcc/12/gcc/aarch64-apple-darwin23/12/include-fixed/objc/runtime.h
  /opt/homebrew/Cellar/gcc@12/12.4.0/bin/../lib/gcc/12/gcc/aarch64-apple-darwin23/12/include-fixed/os/base.h
  /opt/homebrew/Cellar/gcc@12/12.4.0/bin/../lib/gcc/12/gcc/aarch64-apple-darwin23/12/include-fixed/os/trace.h
  /opt/homebrew/Cellar/gcc@12/12.4.0/bin/../lib/gcc/12/gcc/aarch64-apple-darwin23/12/include-fixed/stdint.h
  /opt/homebrew/Cellar/gcc@12/12.4.0/bin/../lib/gcc/12/gcc/aarch64-apple-darwin23/12/include-fixed/stdio.h
  /opt/homebrew/Cellar/gcc@12/12.4.0/bin/../lib/gcc/12/gcc/aarch64-apple-darwin23/12/include-fixed/syslimits.h
```
https://github.com/curl/curl/actions/runs/11732636115/job/32685306324#step:7:49

It's a fairly short list, but still blocks mixing it with any SDK.

Do you think it may gradually reduce to empty?

**Comment by Bo98** [comment]
GCC 15 will clean that up a bit. I know `stdio.h` (`_stdio.h` in macOS 15) still is needed - and is one of the culprits for the incompatiblity between macOS 14 and 15 SDKs. Some of the others will be removed, depending on the OS version you are running. `objc/runtime.h` will still be needed on macOS 13 and 14 for example but will disappear for macOS 15 and later in GCC 15.

I don't have a GCC 15 install to test what the current list looks like on the various macOS versions.

**Comment by vszakats** [comment]
We are possibly seeing a related regression in curl builds, after a recent runner image update (to ):
https://github.com/curl/curl/actions/runs/14378524390/job/40316589059?pr=17012#step:7:169

gcc-13 on macos-15 (with default SDK):
```
configure:5175: gcc-13 -o conftest  --sysroot=/Library/Developer/CommandLineTools/SDKs/MacOSX15.sdk  -w conftest.c  >&5
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX15.sdk/usr/include/_stdio.h:71,
                 from /opt/homebrew/Cellar/gcc@13/13.3.0/lib/gcc/13/gcc/aarch64-apple-darwin24/13/include-fixed/stdio.h:75,
                 from conftest.c:9:
/Library/Developer/CommandLineTools/SDKs/MacOSX15.sdk/usr/include/_stdio.h: In function 'fmemopen':
/Library/Developer/CommandLineTools/SDKs/MacOSX15.sdk/usr/include/_stdio.h:457:107: error: expected declaration specifiers before '__API_AVAILABLE_GET_MACRO_93585900'
  457 | FILE *fmemopen(void * __restrict __buf _LIBC_SIZE(__size), size_t __size, const char * __restrict __mode) __API_AVAILABLE(macos(10.13), ios(11.0), tvos(11.0), watchos(4.0));
      |                                                                                                           ^~~~~~~~~~~~~~~
```

gcc 13.3.0 seems to be the lastest:
```
lrwxr-xr-x  1 runner  admin  23 Apr  8 06:09 gcc@13 -> ../Cellar/gcc@13/13.3.0
```

May be hacklayer issue, or SDK versions going async.

**Comment by Bo98** [comment]
Just GCC 13 or does it affect GCC 14 too?

**Comment by Bo98** [comment]
Ah I see just GCC 13 actually seeing from your link.

If something like `/opt/homebrew/opt/gcc@13/libexec/gcc/aarch64-apple-darwin24/13/install-tools/mkheaders` (as a workaround) doesn't do anything then it might be a deeper compiler issue.

**Comment by vszakats** [comment]
@Bo98 Thanks a lot for this, regenerating the hacklayer solved it: https://github.com/curl/curl/pull/17017

Do you think this is something to fix in the Homebrew package, or maybe some place else?

**Comment by Bo98** [comment]
We're doing CI upgrades early next week and then we can probably rebuild GCC 13 using the new SDK.

Not entirely clear why it only affects GCC 13 though - but I'll check up on that.

**Comment by saschasc** [comment]
Just an assumption:
Eventually this was never backported to gcc 13.x?
https://github.com/iains/gcc-14-branch/commit/843342fa4391c163bb6522839317d81f01c035de

This issue could also help for more context:
https://github.com/iains/gcc-14-branch/issues/12
